File: mopro-r0-example-app/flutter/mopro_flutter_plugin/ios/Classes/MoproFlutterPlugin.swift
import Flutter
import Foundation
import UIKit

class FlutterG1 {
  let x: String
  let y: String
  let z: String

  init(x: String, y: String, z: String) {
    self.x = x
    self.y = y
    self.z = z
  }
}

class FlutterG2 {
  let x: [String]
  let y: [String]
  let z: [String]

  init(x: [String], y: [String], z: [String]) {
    self.x = x
    self.y = y
    self.z = z
  }
}

class FlutterCircomProof {
  let a: FlutterG1
  let b: FlutterG2
  let c: FlutterG1
  let `protocol`: String
  let curve: String

  init(a: FlutterG1, b: FlutterG2, c: FlutterG1, `protocol`: String, curve: String) {
    self.a = a
    self.b = b
    self.c = c
    self.`protocol` = `protocol`
    self.curve = curve
  }
}

class FlutterCircomProofResult {
  let proof: FlutterCircomProof
  let inputs: [String]

  init(proof: FlutterCircomProof, inputs: [String]) {
    self.proof = proof
    self.inputs = inputs
  }
}

func convertCircomProof(res: CircomProofResult) -> [String: Any] {
  let g1a = FlutterG1(x: res.proof.a.x, y: res.proof.a.y, z: res.proof.a.z)
  let g2b = FlutterG2(x: res.proof.b.x, y: res.proof.b.y, z: res.proof.b.z)
  let g1c = FlutterG1(x: res.proof.c.x, y: res.proof.c.y, z: res.proof.c.z)
  let circomProof = FlutterCircomProof(
    a: g1a, b: g2b, c: g1c, `protocol`: res.proof.protocol, curve: res.proof.curve)
  let circomProofResult = FlutterCircomProofResult(proof: circomProof, inputs: res.inputs)
  let resultMap: [String: Any] = [
    "proof": [
      "a": [
        "x": circomProofResult.proof.a.x,
        "y": circomProofResult.proof.a.y,
        "z": circomProofResult.proof.a.z,
      ],
      "b": [
        "x": circomProofResult.proof.b.x,
        "y": circomProofResult.proof.b.y,
        "z": circomProofResult.proof.b.z,
      ],
      "c": [
        "x": circomProofResult.proof.c.x,
        "y": circomProofResult.proof.c.y,
        "z": circomProofResult.proof.c.z,
      ],
      "protocol": circomProofResult.proof.protocol,
      "curve": circomProofResult.proof.curve,
    ],
    "inputs": circomProofResult.inputs,
  ]
  return resultMap
}

func convertCircomProofResult(proof: [String: Any]) -> CircomProofResult {
  let proofMap = proof["proof"] as! [String: Any]
  let aMap = proofMap["a"] as! [String: String]
  let g1a = G1(x: aMap["x"] ?? "0", y: aMap["y"] ?? "0", z: aMap["z"] ?? "1")
  let bMap = proofMap["b"] as! [String: [String]]
  let g2b = G2(
    x: bMap["x"] ?? ["1", "0"], y: bMap["y"] ?? ["1", "0"], z: bMap["z"] ?? ["1", "0"])
  let cMap = proofMap["c"] as! [String: String]
  let g1c = G1(x: cMap["x"] ?? "0", y: cMap["y"] ?? "0", z: cMap["z"] ?? "1")
  let circomProof = CircomProof(
    a: g1a, b: g2b, c: g1c, `protocol`: proofMap["protocol"] as! String,
    curve: proofMap["curve"] as! String)
  let circomProofResult = CircomProofResult(
    proof: circomProof, inputs: proof["inputs"] as! [String])
  return circomProofResult
}

public class MoproFlutterPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "mopro_flutter", binaryMessenger: registrar.messenger())
    let instance = MoproFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "generateCircomProof":
      guard let args = call.arguments as? [String: Any],
        let zkeyPath = args["zkeyPath"] as? String,
        let inputs = args["inputs"] as? String,
        let proofLib = args["proofLib"] as? Int
      else {
        result(
          FlutterError(
            code: "ARGUMENT_ERROR", message: "Missing arguments \(call.arguments)", details: nil))
        return
      }

      do {
        var moproProofLib: ProofLib
        if proofLib == 0 {
          moproProofLib = ProofLib.arkworks
        } else {
          moproProofLib = ProofLib.rapidsnark
        }
        // Call the function from mopro.swift
        let proofResult = try generateCircomProof(
          zkeyPath: zkeyPath, circuitInputs: inputs, proofLib: moproProofLib)
        let resultMap = convertCircomProof(res: proofResult)

        // Return the proof and inputs as a map supported by the StandardMethodCodec
        result(resultMap)
      } catch {
        result(
          FlutterError(
            code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof",
            details: error.localizedDescription))
      }

    case "verifyCircomProof":
      guard let args = call.arguments as? [String: Any],
        let zkeyPath = args["zkeyPath"] as? String,
        let proof = args["proof"] as? [String: Any],
        let proofLib = args["proofLib"] as? Int
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments \(call.arguments)", details: nil))
        return
      }

      do {
        var moproProofLib: ProofLib
        if proofLib == 0 {
          moproProofLib = ProofLib.arkworks
        } else {
          moproProofLib = ProofLib.rapidsnark
        }
        let circomProofResult = convertCircomProofResult(proof: proof)
        // Call the function from mopro.swift
        let valid = try verifyCircomProof(
          zkeyPath: zkeyPath, proofResult: circomProofResult, proofLib: moproProofLib)

        // Return the proof and inputs as a map supported by the StandardMethodCodec
        result(valid)
      } catch {
        result(
          FlutterError(
            code: "PROOF_VERIFICATION_ERROR", message: "Failed to verify proof",
            details: error.localizedDescription))
      }

    case "generateHalo2Proof":
      guard let args = call.arguments as? [String: Any],
        let srsPath = args["srsPath"] as? String,
        let pkPath = args["pkPath"] as? String,
        let inputs = args["inputs"] as? [String: [String]]
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        let proofResult = try generateHalo2Proof(
          srsPath: srsPath, pkPath: pkPath, circuitInputs: inputs)
        let resultMap = [
          "proof": proofResult.proof,
          "inputs": proofResult.inputs,
        ]
        result(resultMap)
      } catch {
        result(
          FlutterError(
            code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof",
            details: error.localizedDescription))
      }

    case "verifyHalo2Proof":
      guard let args = call.arguments as? [String: Any],
        let srsPath = args["srsPath"] as? String
      else {
        result(
          FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments srsPath", details: nil))
        return
      }

      guard let args = call.arguments as? [String: Any],
        let vkPath = args["vkPath"] as? String
      else {
        result(
          FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments vkPath", details: nil))
        return
      }

      guard let args = call.arguments as? [String: Any],
        let proof = args["proof"] as? FlutterStandardTypedData
      else {
        result(
          FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments proof", details: nil))
        return
      }

      guard let args = call.arguments as? [String: Any],
        let inputs = args["inputs"] as? FlutterStandardTypedData
      else {
        result(
          FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments inputs", details: nil))
        return
      }

      do {
        let valid = try verifyHalo2Proof(
          srsPath: srsPath, vkPath: vkPath, proof: proof.data, publicInput: inputs.data)
        result(valid)
      } catch {
        result(
          FlutterError(
            code: "PROOF_VERIFICATION_ERROR", message: "Failed to verify proof",
            details: error.localizedDescription))
      }
    case "generateNoirProof":
      guard let args = call.arguments as? [String: Any],
        let circuitPath = args["circuitPath"] as? String,
        let inputs = args["inputs"] as? [String],
        let onChain = args["onChain"] as? Bool,
        let vk = args["vk"] as? FlutterStandardTypedData,
        let lowMemoryMode = args["lowMemoryMode"] as? Bool
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      let srsPath = args["srsPath"] as? String

      do {
        let proofResult = try generateNoirProof(
          circuitPath: circuitPath, srsPath: srsPath, inputs: inputs, onChain: onChain, vk: vk.data, lowMemoryMode: lowMemoryMode)
        result(proofResult)
      } catch {
        result(
          FlutterError(
            code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof",
            details: error.localizedDescription))
      }

    case "verifyNoirProof":
      guard let args = call.arguments as? [String: Any],
        let circuitPath = args["circuitPath"] as? String,
        let proof = args["proof"] as? FlutterStandardTypedData,
        let onChain = args["onChain"] as? Bool,
        let vk = args["vk"] as? FlutterStandardTypedData,
        let lowMemoryMode = args["lowMemoryMode"] as? Bool
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        let valid = try verifyNoirProof(circuitPath: circuitPath, proof: proof.data, onChain: onChain, vk: vk.data, lowMemoryMode: lowMemoryMode)
        result(valid)
      } catch {
        result(
          FlutterError(
            code: "PROOF_VERIFICATION_ERROR", message: "Failed to verify proof",
            details: error.localizedDescription))
      }
    case "getNoirVerificationKey":
      guard let args = call.arguments as? [String: Any],
        let circuitPath = args["circuitPath"] as? String,
        let onChain = args["onChain"] as? Bool,
        let lowMemoryMode = args["lowMemoryMode"] as? Bool
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      let srsPath = args["srsPath"] as? String

      do {
        let vkResult = try getNoirVerificationKey(circuitPath: circuitPath, srsPath: srsPath, onChain: onChain, lowMemoryMode: lowMemoryMode)
        result(vkResult)
      } catch {
        result(
          FlutterError(
            code: "VK_GENERATION_ERROR", message: "Failed to generate verification key",
            details: error.localizedDescription))
      }
    case "generateRisc0Proof":
      guard let args = call.arguments as? [String: Any],
        let input = args["input"] as? Int
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        let proofResult = try risc0Prove(input: UInt32(input))
        let resultMap: [String: Any] = [
          "receipt": proofResult.receipt,
        ]
        result(resultMap)
      } catch {
        result(
          FlutterError(
            code: "PROOF_GENERATION_ERROR", message: "Failed to generate RISC0 proof",
            details: error.localizedDescription))
      }
    case "verifyRisc0Proof":
      guard let args = call.arguments as? [String: Any],
        let receiptBytes = args["receiptBytes"] as? FlutterStandardTypedData
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        let verifyResult = try risc0Verify(receiptBytes: receiptBytes.data)
        let resultMap: [String: Any] = [
          "isValid": verifyResult.isValid,
          "outputValue": verifyResult.outputValue,
        ]
        result(resultMap)
      } catch {
        result(
          FlutterError(
            code: "PROOF_VERIFICATION_ERROR", message: "Failed to verify RISC0 proof",
            details: error.localizedDescription))
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: mopro-r0-example-app/flutter/mopro_flutter_plugin/android/src/main/kotlin/com/example/mopro_flutter/MoproFlutterPlugin.kt
package com.example.mopro_flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import uniffi.MoproR0ExampleApp.*

import io.flutter.plugin.common.StandardMethodCodec

class FlutterG1(x: String, y: String, z: String) {
    val x = x
    val y = y
    val z = z
}

class FlutterG2(x: List<String>, y: List<String>, z: List<String>) {
    val x = x
    val y = y
    val z = z
}

class FlutterCircomProof(a: FlutterG1, b: FlutterG2, c: FlutterG1, protocol: String, curve: String) {
    val a = a
    val b = b
    val c = c
    val protocol = protocol
    val curve = curve
}

class FlutterCircomProofResult(proof: FlutterCircomProof, inputs: List<String>) {
    val proof = proof
    val inputs = inputs
}

fun convertCircomProof(res: CircomProofResult): Map<String, Any> {
    val g1a = FlutterG1(res.proof.a.x, res.proof.a.y, res.proof.a.z)
            val g2b = FlutterG2(res.proof.b.x, res.proof.b.y, res.proof.b.z)
            val g1c = FlutterG1(res.proof.c.x, res.proof.c.y, res.proof.c.z)
            val circomProof = FlutterCircomProof(g1a, g2b, g1c, res.proof.protocol, res.proof.curve)
            val circomProofResult = FlutterCircomProofResult(circomProof, res.inputs)
            // Convert to Map before sending
    val resultMap = mapOf(
        "proof" to mapOf(
            "a" to mapOf(
                "x" to circomProofResult.proof.a.x,
                "y" to circomProofResult.proof.a.y,
                "z" to circomProofResult.proof.a.z
            ),
            "b" to mapOf(
                "x" to circomProofResult.proof.b.x,
                "y" to circomProofResult.proof.b.y,
                "z" to circomProofResult.proof.b.z
            ),
            "c" to mapOf(
                "x" to circomProofResult.proof.c.x,
                "y" to circomProofResult.proof.c.y,
                "z" to circomProofResult.proof.c.z
            ),
            "protocol" to circomProofResult.proof.protocol,
            "curve" to circomProofResult.proof.curve
        ),
        "inputs" to circomProofResult.inputs
    )
    return resultMap
}
fun convertCircomProofResult(proofResult: Map<String, Any>): CircomProofResult {
    val proofMap = proofResult["proof"] as Map<String, Any>
    val aMap = proofMap["a"] as Map<String, Any>
    val g1a = G1(
        aMap["x"] as String,
        aMap["y"] as String,
        aMap["z"] as String
    )
    val bMap = proofMap["b"] as Map<String, Any>
    val g2b = G2(
        bMap["x"] as List<String>,
        bMap["y"] as List<String>,
        bMap["z"] as List<String>
    )
    val cMap = proofMap["c"] as Map<String, Any>
    val g1c = G1(
        cMap["x"] as String,
        cMap["y"] as String,
        cMap["z"] as String
    )
    val circomProof = CircomProof(
        g1a,
        g2b,
        g1c,
        proofMap["protocol"] as String,
        proofMap["curve"] as String
    )
    val circomProofResult = CircomProofResult(circomProof, proofResult["inputs"] as List<String>)
    return circomProofResult
  }

/** MoproFlutterPlugin */
class MoproFlutterPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "mopro_flutter",
            StandardMethodCodec.INSTANCE
        )
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "generateCircomProof") {
            val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing zkeyPath",
                null
            )

            val inputs =
                call.argument<String>("inputs") ?: return result.error(
                    "ARGUMENT_ERROR",
                    "Missing inputs",
                    null
                )
            
            val proofLibIndex = call.argument<Int>("proofLib") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proofLib",
                null
            )

            val proofLib = if (proofLibIndex == 0) ProofLib.ARKWORKS else ProofLib.RAPIDSNARK

            val res = generateCircomProof(zkeyPath, inputs, proofLib)
            val resultMap = convertCircomProof(res)

            
            result.success(resultMap)
        } else if (call.method == "verifyCircomProof") {
            val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing zkeyPath",
                null
            )

            val proof = call.argument<Map<String, Any>>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )

            val proofLibIndex = call.argument<Int>("proofLib") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proofLib",
                null
            )

            val proofLib = if (proofLibIndex == 0) ProofLib.ARKWORKS else ProofLib.RAPIDSNARK

            val circomProofResult = convertCircomProofResult(proof)
            val res = verifyCircomProof(zkeyPath, circomProofResult, proofLib)
            result.success(res)

        } else if (call.method== "generateHalo2Proof") {
            val srsPath = call.argument<String>("srsPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing srsPath",
                null
            )

            val pkPath = call.argument<String>("pkPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing pkPath",
                null
            )

            val inputs = call.argument<Map<String, List<String>>>("inputs") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputs",
                null
            )

            val res = generateHalo2Proof(srsPath, pkPath, inputs)
            val resultMap = mapOf(
                "proof" to res.proof,
                "inputs" to res.inputs
            )

            result.success(resultMap)
        } else if (call.method== "verifyHalo2Proof") {
            val srsPath = call.argument<String>("srsPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing srsPath",
                null
            )

            val vkPath = call.argument<String>("vkPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing vkPath",
                null
            )

            val proof = call.argument<ByteArray>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )

            val inputs = call.argument<ByteArray>("inputs") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputs",
                null
            )

            val res = verifyHalo2Proof(srsPath, vkPath, proof, inputs)
            result.success(res)
        } else if (call.method== "generateNoirProof") {
            val circuitPath = call.argument<String>("circuitPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing circuitPath",
                null
            )

            val srsPath = call.argument<String>("srsPath") 

            val inputs = call.argument<List<String>>("inputs") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputs",
                null
            )

            val onChain = call.argument<Boolean>("onChain") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing onChain",
                null
            )

            val vk = call.argument<ByteArray>("vk") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing vk",
                null
            )

            val lowMemoryMode = call.argument<Boolean>("lowMemoryMode") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing lowMemoryMode",
                null
            )

            val res = generateNoirProof(circuitPath, srsPath, inputs, onChain, vk, lowMemoryMode)
            result.success(res)
        } else if (call.method== "verifyNoirProof") {
            val circuitPath = call.argument<String>("circuitPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing circuitPath",
                null
            )

            val proof = call.argument<ByteArray>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )

            val onChain = call.argument<Boolean>("onChain") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing onChain",
                null
            )

            val vk = call.argument<ByteArray>("vk") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing vk",
                null
            )

            val lowMemoryMode = call.argument<Boolean>("lowMemoryMode") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing lowMemoryMode",
                null
            )

            val res = verifyNoirProof(circuitPath, proof, onChain, vk, lowMemoryMode)
            result.success(res)

        } else if (call.method== "getNoirVerificationKey") {
            val circuitPath = call.argument<String>("circuitPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing circuitPath",
                null
            )

            val srsPath = call.argument<String>("srsPath")

            val onChain = call.argument<Boolean>("onChain") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing onChain",
                null
            )

            val lowMemoryMode = call.argument<Boolean>("lowMemoryMode") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing lowMemoryMode",
                null
            )

            val res = getNoirVerificationKey(circuitPath, srsPath, onChain, lowMemoryMode)
            result.success(res)

        } else if (call.method== "generateRisc0Proof") {
            val input = call.argument<Int>("input") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing input",
                null
            )

            try {
                val res = risc0Prove(input.toUInt())
                val resultMap = mapOf(
                    "receipt" to res.receipt
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("PROOF_GENERATION_ERROR", "Failed to generate RISC0 proof", e.message)
            }

        } else if (call.method== "verifyRisc0Proof") {
            val receiptBytes = call.argument<ByteArray>("receiptBytes") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing receiptBytes",
                null
            )

            try {
                val res = risc0Verify(receiptBytes)
                val resultMap = mapOf(
                    "isValid" to res.isValid,
                    "outputValue" to res.outputValue.toInt()
                )
                result.success(resultMap)
            } catch (e: Exception) {
                result.error("PROOF_VERIFICATION_ERROR", "Failed to verify RISC0 proof", e.message)
            }

        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
